src/main/java/net/fabricmc/loom/task/MigrateMappingsTask.java

```diff
@@ -61,123 +68,130 @@ public void doTask() throws Throwable {
+			throw new IllegalArgumentException("Could not find input directory: " + inputDirPath.toAbsolutePath());
 		}
 
-		if (!outputDir.exists()) {
-			if (!outputDir.mkdirs()) {
-				throw new RuntimeException("Could not create output directory:" + outputDir);
-			}
-		}
+		Files.createDirectories(outputDirPath);
 
-		Mappings sourceMappings = extension.getMappingsProvider().getMappings();
-		Mappings targetMappings;
+		MappingsProvider mappingsProvider = extension.getMappingsProvider();
 
-		try (FileInputStream stream = new FileInputStream(mappingsFile)) {
-			targetMappings = net.fabricmc.mappings.MappingsProvider.readTinyMappings(stream, false);
+		try {
+			TinyTree currentMappings = mappingsProvider.getMappings();
+			TinyTree targetMappings = mappingsProvider.getMappingsOfVersion(project, targetMappingsVersion);
+			migrateMappings(project, extension.getMinecraftMappedProvider(), inputDirPath, outputDirPath, currentMappings, targetMappings, extension);
+			project.getLogger().lifecycle(":remapped project written to " + outputDirPath.toAbsolutePath());
+		} catch (IOException e) {
+			throw new IllegalArgumentException("Could not find mappings for version " + targetMappingsVersion, e);
 		}
+	}
 
+	private static void migrateMappings(Project project, MinecraftMappedProvider minecraftMappedProvider,
+										Path inputDir, Path outputDir, TinyTree currentMappings, TinyTree targetMappings, LoomGradleExtension extension
+	) throws IOException {
 		project.getLogger().lifecycle(":joining mappings");
-		MappingSet mappingSet = new MappingsJoiner(sourceMappings, targetMappings, "intermediary", "named").read();
+		MappingSet mappingSet = new MappingsJoiner(currentMappings, targetMappings,
+						"intermediary", "named").read();
 
 		project.getLogger().lifecycle(":remapping");
-		Mercury mercury = new Mercury();
+		Mercury mercury = SourceRemapper.createMercuryWithClassPath(project, false);
 
-		for (File file : project.getConfigurations().getByName(Constants.MINECRAFT_DEPENDENCIES).getFiles()) {
-			mercury.getClassPath().add(file.toPath());
-		}
-
-		for (File file : project.getConfigurations().getByName("compileClasspath").getFiles()) {
-			mercury.getClassPath().add(file.toPath());
-		}
+		mercury.getClassPath().add(minecraftMappedProvider.MINECRAFT_MAPPED_JAR.toPath());
+		mercury.getClassPath().add(minecraftMappedProvider.MINECRAFT_INTERMEDIARY_JAR.toPath());
 
 		mercury.getClassPath().add(extension.getMinecraftMappedProvider().MINECRAFT_MAPPED_JAR.toPath());
 		mercury.getClassPath().add(extension.getMinecraftMappedProvider().MINECRAFT_INTERMEDIARY_JAR.toPath());
 
-		mercury.getProcessors().add(MercuryRemapper.create(mappingSet));
-
 		try {
-			mercury.rewrite(inputDir.toPath(), outputDir.toPath());
+			mercury.rewrite(inputDir, outputDir);
 		} catch (Exception e) {
 			project.getLogger().warn("Could not remap fully!", e);
 		}
 
-		project.getLogger().lifecycle(":cleaning file descriptors");
-		System.gc();
```

Comment:
This needs adding back as much as we all hate it. It mostly fixes a bug with the Eclipse AST holding onto files.

Reply:
No comment about how I have the same block of code twice for no reason!